### PR TITLE
[tensorpipe] create a new port

### PR DESCRIPTION
**Pull request by luncliff**
### What does your PR fix?

There was no port request for this project.

This is one of the 3rd party libraries for [the PyTorch project](https://github.com/pytorch/pytorch/blob/master/.gitmodules). The PR will be used for future support of the `libtorch` port.

### Which triplets are supported/not supported? Have you updated the CI baseline?

According to the project's CI configuration, its target is `linux`, `osx`. And there are no sources for `windows`, `uwp` triplets. 

* `x64-osx`
* `x64-linux`

### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?

* Make separate Pull Requests per port. **Requires #16471**
* `"version-date"` instead of `"version-string"`, [formatted as `YYYY-MM-DD`](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/manifest-files.md#version-fields).

### CODE REVIEW

**Comment by luncliff** [comment]
The port will be rebased when #16471 is merged. To prevent merge, it will remain in the draft mode.

File: ports/tensorpipe/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+  FEATURES
+    cuda        TP_USE_CUDA
+    cuda        TP_ENABLE_CUDA_IPC
+    shm         TP_ENABLE_SHM
+    ibv         TP_ENABLE_IBV
+    cma         TP_ENABLE_CMA
+    pybind11    TP_BUILD_PYTHON
+)
```
Comment (NancyLi1013) [inline ports/tensorpipe/portfile.cmake:21]:
```suggestion
vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
    FEATURES
        cuda        TP_USE_CUDA
        cuda        TP_ENABLE_CUDA_IPC
        shm         TP_ENABLE_SHM
        ibv         TP_ENABLE_IBV
        cma         TP_ENABLE_CMA
        pybind11    TP_BUILD_PYTHON
)
```
There should be 4 spaces.

File: ports/tensorpipe/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS
+        ${FEATURE_OPTIONS}
+        -DTP_BUILD_TESTING=OFF -DTP_BUILD_BENCHMARK=OFF -DTP_BUILD_LIBUV=OFF
```
Comment (NancyLi1013) [inline ports/tensorpipe/portfile.cmake:28]:
```suggestion
        -DTP_BUILD_TESTING=OFF 
        -DTP_BUILD_BENCHMARK=OFF 
        -DTP_BUILD_LIBUV=OFF
```

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
Hi @luncliff 

Thanks for your PR.

Have you tested these features?

**Comment by luncliff** [comment]
@NancyLi1013 not yet. Doing so requires a Linux device with CUDA but I don't have one now. Asking my friends now :(

**Comment by luncliff** [comment]
I'd like to try `vcpkg_cmake_configure` and `vcpkg_cmake_install` for this port. Any suggestions?

**Comment by NancyLi1013** [comment]
> I'd like to try [`vcpkg_cmake_configure`](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/ports/vcpkg-cmake/vcpkg_cmake_configure.md) and [`vcpkg_cmake_install`](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/ports/vcpkg-cmake/vcpkg_cmake_install.md) for this port. Any suggestions?

Currently, there is no need to do any actions for the function changes. Just keep them as before. Maybe we will update them to new functions in the near future, but not now.

File: ports/tensorpipe/portfile.cmake [hunk 3]
```diff
@@ -18,15 +18,15 @@ vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
         ibv         TP_ENABLE_IBV
         cma         TP_ENABLE_CMA
         pybind11    TP_BUILD_PYTHON
+        test        TP_BUILD_TESTING
+        test        TP_BUILD_BENCHMARK
```
Comment (NancyLi1013) [inline ports/tensorpipe/portfile.cmake:22]: This might be feature `benchmark` instead of `test`.

` benchmark           TP_BUILD_BENCHMARK`

So you also need to add another feature `benchmark` in vcpkg.json.

File: ports/tensorpipe/fix-cmakelists.patch [hunk 4]
```diff
@@ -43,3 +58,37 @@ index 0a0944b..b7ff4fa 100644
...
+-
++find_package(GTest CONFIG REQUIRED)
+ target_link_libraries(tensorpipe_test PRIVATE
+   tensorpipe
+-  uv::uv
+-  gmock
+-  gtest_main)
++  unofficial::libuv::libuv
++  GTest::gmock GTest::gtest_main
++)
```
Comment (NancyLi1013) [inline ports/tensorpipe/fix-cmakelists.patch:94]: I prefer to make an separate patch for the feature. What do you think?

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: ports/tensorpipe/fix-cmakelists.patch (hunk 4 above)
Comment (luncliff) [reply ports/tensorpipe/fix-cmakelists.patch:94]: Why not? I will separate the part

File: ports/tensorpipe/portfile.cmake (hunk 3 above)
Comment (luncliff) [reply ports/tensorpipe/portfile.cmake:22]: Got it. I will update the part when I get a Linux machine.

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
Hi @luncliff 

Could you please help confirm the test status of features?

**Comment by luncliff** [comment]
Sadly, the `tensorpipe_test` keeps failing. for some features...

* `pybind11` needs an additional patch for its build. I will update the `fix-cmakelists.patch`
* `cma`, `ibv`, `cuda`: build passed, but `tensorpipe_test` failes with infinite loop
* `shm`, `benchmark`: seems like these are usable

Does every `test` executable in the port must report green? 🤔 If so, I will change this to draft again then check the mainstream's CI.

**Comment by NancyLi1013** [comment]
Generally, we should keep all features work normally for every port. If one or more of these features cannot work. it would be better to fix them. If they can only support the specific platform, we should post some messages to describe it. Otherwise, you can choose to remove them from feature lists until they can work fine.

**Comment by luncliff** [comment]
I see. I'd better remove `cma` and `ibv` which doesn't pass the `test`. I want to support `cuda`, but it may not that easy to debug with current state.. :( 
I will try more after fixing `pybind11` related works.

**Comment by luncliff** [comment]
After 97bb90b, `tensorpipe`'s CUDA test passes.

```console
user@host:~/vcpkg$ ./vcpkg install tensorpipe[cuda,test]
Computing installation plan...
The following packages will be built and installed:
    tensorpipe[core,cuda,test]:x64-linux -> 2021-04-26
Detecting compiler hash for triplet x64-linux...
Could not locate cached archive: /home/luncliff/.cache/vcpkg/archives/5f/5fbca65575504f45776aebee0cf1f263fa52ac5d.zip
Starting package 1/1: tensorpipe:x64-linux
Building package tensorpipe[core,cuda,test]:x64-linux...
-- Using cached /home/luncliff/vcpkg/downloads/pytorch-tensorpipe-c5a21994bc766659f7f85edb75478e13f429f46c.tar.gz
-- Cleaning sources at /home/luncliff/vcpkg/buildtrees/tensorpipe/src/13f429f46c-d111f7299b.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source /home/luncliff/vcpkg/downloads/pytorch-tensorpipe-c5a21994bc766659f7f85edb75478e13f429f46c.tar.gz
-- Applying patch fix-cmakelists.patch
-- Applying patch support-test.patch
-- Applying patch support-pybind11.patch
-- Using source at /home/luncliff/vcpkg/buildtrees/tensorpipe/src/13f429f46c-d111f7299b.clean
...
Installing package tensorpipe[core,cuda,test]:x64-linux... done
Elapsed time for package tensorpipe:x64-linux: 1.606 min

Total elapsed time: 1.612 min

The package tensorpipe:x64-linux provides CMake targets:

    find_package(tensorpipe CONFIG REQUIRED)
    target_link_libraries(main PRIVATE tensorpipe)

```

```console
user@host:~/vcpkg$ pushd ./buildtrees/tensorpipe/x64-linux-dbg/tensorpipe/test/
~/vcpkg/buildtrees/tensorpipe/x64-linux-dbg/tensorpipe/test ~/vcpkg
user@host:~/vcpkg/buildtrees/tensorpipe/x64-linux-dbg/tensorpipe/test$ ./tensorpipe_test
...
[  SKIPPED ] CudaIpc/CudaMultiGPUChannelTestSuite.SendReverseAcrossDevices/0
[  SKIPPED ] CudaIpc/CudaMultiGPUChannelTestSuite.SendAcrossNonDefaultDevices/0

  YOU HAVE 4 DISABLED TESTS

```

**Comment by luncliff** [comment]
For `x64-osx` failure, this will be rebased after #17610

**Comment by luncliff** [comment]
Test execution in Darwin also passed

```console
user@host$ pushd /Users/user/dev/vcpkg/buildtrees/tensorpipe/x64-osx-dbg/tensorpipe/test
user@host$ ./tensorpipe_test 
[----------] 1 test from Mpt/MptChannelTestSuite
[ RUN      ] Mpt/MptChannelTestSuite.ContextIsNotJoined/0
[       OK ] Mpt/MptChannelTestSuite.ContextIsNotJoined/0 (1 ms)
[----------] 1 test from Mpt/MptChannelTestSuite (1 ms total)

[----------] Global test environment tear-down
[==========] 52 tests from 15 test suites ran. (6289 ms total)
[  PASSED  ] 52 tests.

  YOU HAVE 2 DISABLED TESTS


```

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
LGTM now, thanks for adding this port @luncliff.
